addons/controllers/clusterbootstrap_controller.go

```diff
@@ -1464,3 +1539,157 @@ func (r *ClusterBootstrapReconciler) reconcileClusterProxyAndNetworkSettings(clu
 
 	return nil
 }
+
+func (r *ClusterBootstrapReconciler) clusterIsReadyForDeletion(cluster *clusterapiv1beta1.Cluster, log logr.Logger) (bool, error) {
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	err := r.Client.Get(r.context, client.ObjectKeyFromObject(cluster), clusterBootstrap)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "failed to lookup clusterbootstrap")
+		return false, err
+	}
+	if apierrors.IsNotFound(err) {
+		log.Info("cluster bootstrap not found")
+	} else {
+		remoteClient, err := util.GetClusterClient(r.context, r.Client, r.Scheme, clusterapiutil.ObjectKey(cluster))
+		if err != nil {
+			return false, err
+		}
+		if hasPackageInstalls(r.context, remoteClient, cluster, r.Config.SystemNamespace, clusterBootstrap.Spec.AdditionalPackages, log) {
+			err = r.removeAdditionalPackageInstalls(remoteClient, cluster, clusterBootstrap, log)
+			if err != nil {
+				return false, err
+			}
+			return false, nil
+		}
+	}
+	log.Info("cluster is ready for deletion")
+	return true, nil
+}
+
+func (r *ClusterBootstrapReconciler) removeFinalizersFromClusterResources(cluster *clusterapiv1beta1.Cluster, log logr.Logger) error {
+	log.Info("removing finalizer from clusterbootstrap if present")
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	err := r.Client.Get(r.context, client.ObjectKeyFromObject(cluster), clusterBootstrap)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "failed to lookup clusterbootstrap")
+		return err
+	}
+	if !apierrors.IsNotFound(err) {
+		err = r.removeFinalizer(clusterBootstrap)
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Info("removing finalizer from cluster kubeconfig secret if present")
+	clusterKubeConfigSecret := &corev1.Secret{} // TODO (adduarte): test without deleting clusterKubeConfig
+	key := client.ObjectKey{Namespace: cluster.Namespace, Name: secretutil.Name(cluster.Name, secretutil.Kubeconfig)}
+	err = r.Client.Get(r.context, key, clusterKubeConfigSecret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "unable to fetch cluster kubeconfig secret")
+		return err
+	}
+	if !apierrors.IsNotFound(err) {
+		err = r.removeFinalizer(clusterKubeConfigSecret)
+		if err != nil {
+			return err
+		}
+	}
+
+	// remove finalizer from cluster
+	log.Info("removing finalizer from cluster if present")
+	err = r.removeFinalizer(cluster)
+	if err != nil {
+		return err
+	}
+
+	log.Info("all finalizers have been removed from cluster resources")
+	return nil
+}
+
+func (r *ClusterBootstrapReconciler) reconcileDelete(cluster *clusterapiv1beta1.Cluster, log logr.Logger) (ctrl.Result, error) {
+	okToRemoveFinalizers := false
+	var err error
+
+	timeOutReached := time.Now().After(cluster.GetDeletionTimestamp().Add(r.Config.ClusterDeleteTimeout))
+	if timeOutReached {
+		okToRemoveFinalizers = true
+	} else {
+		okToRemoveFinalizers, err = r.clusterIsReadyForDeletion(cluster, log)
```

Comment:
Nit: Looks like we are checking for deletion status, and doing the actual deletion in this function as well. However the function name suggests it's only checking for deletion status, could use a better name.

Reply:
renamed to makeClusterReadyForDeletion